File: pulsar-skeleton/src/main/java/ai/platon/pulsar/crawl/fetch/FetchTask.java
package ai.platon.pulsar.crawl.fetch;

import ai.platon.pulsar.common.URLUtil;
import ai.platon.pulsar.common.Urls;
import ai.platon.pulsar.persist.WebPage;

import javax.annotation.Nonnull;
import java.net.URL;
import java.time.Instant;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * This class described the item to be fetched.
 */
public class FetchTask implements Comparable<FetchTask> {

    /**
     * The initial value is the current timestamp in second, to make it
     * an unique id in the current host
     */
    private static AtomicInteger instanceSequence = new AtomicInteger(0);
    private Key key;
    private WebPage page;
    private Instant pendingStart = Instant.EPOCH;

    public FetchTask(int jobID, int priority, String protocol, String host, WebPage page, URL u) {
        this.key = new Key(jobID, priority, protocol, host, u.toString());
        this.page = page;
    }

    /**
     * Create an item. Queue id will be created based on <code>groupMode</code>
     * argument, either as a protocol + hostname pair, protocol + IP
     * address pair or protocol+domain pair.
     */
    public static FetchTask create(int jobId, int priority, String url, WebPage page, URLUtil.GroupMode groupMode) {
        final URL u = Urls.getURLOrNull(url);

        if (u == null) {
            return null;
        }

        final String proto = u.getProtocol();
        final String host = URLUtil.getHost(u, groupMode);

        if (proto == null || host.isEmpty()) {
            return null;
        }

        return new FetchTask(jobId, priority, proto.toLowerCase(), host.toLowerCase(), page, u);
    }

    public WebPage getPage() {
        return page;
    }

    public void setPage(WebPage page) {
        this.page = page;
    }

    public Key getKey() {
        return key;
    }

    public int getPriority() {
        return key.priority;
    }

    public String getProtocol() {
        return key.protocol;
    }

    public String getHost() {
        return key.host;
    }

    public int getItemId() {
        return key.itemId;
    }

    public String getUrl() {
        return key.url;
    }

    public Instant getPendingStart() {
        return pendingStart;
    }

    public void setPendingStart(Instant pendingStart) {
        this.pendingStart = pendingStart;
    }

    /**
     * Generate an unique numeric id for the fetch item
     * <p>
     * fetch item id is used to easy item searching
     */
    private int nextId() {
        return instanceSequence.incrementAndGet();
    }

    @Override
    public String toString() {
        return "[itemId=" + key.itemId + ", priority=" + key.priority + ", url=" + key.url + "]";
    }

    @Override
    public int compareTo(@Nonnull FetchTask fetchTask) {
        return getKey().compareTo(fetchTask.getKey());
    }

    public class Key implements Comparable<Key> {
        public int jobID;
        public int priority;
        public String protocol;
        public String host;
        public int itemId;
        public String url;

        public Key(int jobID, int priority, String protocol, String host, String url) {
            this.jobID = jobID;
            this.priority = priority;
            this.protocol = protocol;
            this.host = host;
            this.itemId = nextId();
            this.url = url;
        }

        @Override
        public int compareTo(Key key) {
            return itemId - key.itemId;
        }
    }
}

File: pulsar-skeleton/src/main/kotlin/ai/platon/pulsar/crawl/fetch/FetchTask.kt
package ai.platon.pulsar.crawl.fetch

import ai.platon.pulsar.common.URLUtil
import ai.platon.pulsar.common.Urls
import ai.platon.pulsar.crawl.fetch.data.PoolId
import ai.platon.pulsar.persist.WebPage
import java.net.URL
import java.time.Instant
import java.util.concurrent.atomic.AtomicInteger

interface IFetchEntry {
    var reservedUrl: String
    var page: WebPage
}

/**
 * This class described the item to be fetched.
 */
class FetchTask private constructor(
        jobID: Int,
        val priority: Int,
        val protocol: String,
        val host: String,
        val page: WebPage,
        val url: URL
) : Comparable<FetchTask> {
    val key = Key(jobID, priority, protocol, host, url.toString())
    val itemId get() = key.itemId
    val urlString get() = key.url
    val poolId get() = PoolId(priority, protocol, host)
    var pendingStart = Instant.EPOCH!!

    override fun toString(): String {
        return "<itemId=" + key.itemId + ", priority=" + key.priority + ", url=" + key.url + ">"
    }

    override fun compareTo(other: FetchTask): Int {
        return key.compareTo(other.key)
    }

    data class Key(
            var jobID: Int,
            var priority: Int,
            var protocol: String,
            var host: String,
            var url: String,
            var itemId: Int = nextId()
    ) : Comparable<Key> {
        override fun compareTo(other: Key): Int {
            return itemId - other.itemId
        }
    }

    companion object {
        /**
         * The initial value is the current timestamp in second, to make it
         * an unique id in the current host
         */
        private val instanceSequence = AtomicInteger(0)

        /**
         * Generate an unique numeric id for the fetch item
         * fetch item id is used to easy item searching
         */
        private fun nextId(): Int {
            return instanceSequence.incrementAndGet()
        }

        /**
         * Create an item. Queue id will be created based on `groupMode`
         * argument, either as a protocol + hostname pair, protocol + IP
         * address pair or protocol+domain pair.
         */
        fun create(jobId: Int, priority: Int, url: String, page: WebPage, groupMode: URLUtil.GroupMode): FetchTask? {
            val u = Urls.getURLOrNull(url) ?: return null

            val proto = u.protocol
            val host = URLUtil.getHost(u, groupMode)

            return if (proto == null || host.isEmpty()) {
                null
            } else FetchTask(jobId, priority, proto.toLowerCase(), host.toLowerCase(), page, u)
        }
    }
}

/**
 * Created by vincent on 16-10-15.
 * Copyright @ 2013-2016 Platon AI. All rights reserved
 */
data class FetchStat(
        var hostName: String,
        var urls: Int = 0,
        var indexUrls: Int = 0,
        var detailUrls: Int = 0,
        var searchUrls: Int = 0,
        var mediaUrls: Int = 0,
        var bbsUrls: Int = 0,
        var blogUrls: Int = 0,
        var tiebaUrls: Int = 0,
        var unknownUrls: Int = 0,
        var urlsTooLong: Int = 0,
        var urlsFromSeed: Int = 0,
        var cookieView: Int = 0
) : Comparable<FetchStat> {

    override fun compareTo(other: FetchStat): Int {
        val reverseHost = Urls.reverseHost(hostName)
        val reverseHost2 = Urls.reverseHost(other.hostName)

        return reverseHost.compareTo(reverseHost2)
    }
}

data class BatchStat(
        var numTaskDone: Int = 0,
        var numSuccessTasks: Int = 0,
        var numIncompletePages: Int = 0,
        var numFailedTasks: Int = 0,
        var totalBytes: Long = 0L
) {
    val averagePageSize get() = 1.0 * totalBytes / (0.1 + numSuccessTasks)
}
